### PR TITLE
feat: add loot tables

### LINKED ISSUE #1378: [RFC] Loot Tables
**Is your feature request related to a problem? Please describe.**

The Anime Game's key gameplay is about getting resources. To pull gacha, collect materials in the big world, kill monsters or challenging the domains can be viewed as different way to collect different resources. Loot Tables are technical files to grant overall control to every aspect of the "resource collecting" gameplay.

**Describe the solution you'd like**

I have implemented a basic prototype of the Loot Table mechanism and want future discussion and comments. Please see the draft Pull Request for the implementation details.

**Additional context**
The design itself is basically a *clean room reimplementation* of [the same-name concept in Minecraft](https://minecraft.fandom.com/wiki/Loot_table). Its usage should be easily understood with detailed documents in Minecraft community.

### PR DESCRIPTION
**Pull request by diauweb**
## Description
See #1378.
Please note this is a prototype and need future discussions.

Please carefully read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md) before making any pull requests.
And, **Do not make a pull request to merge into stable unless it is a hotfix. Use the development branch instead.**
## Issues fixed by this PR
#1378 

## Type of changes

- [ ] Bug fix
- [x] New feature 
- [x] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

### CODE REVIEW

**Comment by ghost** [comment]
I like your loot table schema, but personally would prefer every monster loot table in one json file instead of in many files (per monster).

**Comment by diauweb** [comment]
> I like your loot table schema, but personally would prefer every monster loot table in one json file instead of in many files (per monster).

I afraid we'll eventually get a very huge file. 
How about organize the tables by monster group? e.g. All types of Hilichurls use one file with conditional tweaks in loot table rules

**Comment by GiRx8** [comment]
> > I like your loot table schema, but personally would prefer every monster loot table in one json file instead of in many files (per monster).
> 
> I afraid we'll eventually get a very huge file. How about organize the tables by monster group? e.g. All types of Hilichurls use one file with conditional tweaks in loot table rules

I agree, it would become too messy after a while. By mob type / groups within table would be easier to maintain.

**Comment by ghost** [comment]
> I agree, it would become too messy after a while. By mob type / groups within table would be easier to maintain.

The reason i prefer it to be in one file is because i would rather not have to deal with 100s of json files with one for each monster. It would be better to make a tool to create drop tables for each monster and then have it compile everything into one json.

**Comment by GiRx8** [comment]
> > I agree, it would become too messy after a while. By mob type / groups within table would be easier to maintain.
> 
> The reason i prefer it to be in one file is because i would rather not have to deal with 100s of json files with one for each monster. It would be better to make a tool to create drop tables for each monster and then have it compile everything into one json.

Yeah big + for a tool, but yeah as it stands it’s a huge mess. Will be good to organize it at least.

**Comment by diauweb** [comment]
I added a new layer `LootRegistry` which can help reduce the count of the files

**Comment by diauweb** [comment]
I have added an example on handling gacha with loot table

**Comment by diauweb** [comment]
I'll mark this pr as ready as I have implemented most of it.
Future developments and migrations can be done in other new prs.

File: src/main/java/emu/grasscutter/game/gacha/GachaManager.java [hunk 1]
```diff
@@ -296,24 +299,34 @@ public synchronized void doPulls(Player player, int scheduleId, int times) {
...
-			ItemData itemData = GameData.getItemDataMap().get(itemId);
-			if (itemData == null) {
+            int itemId;
+            ItemData itemData;
+            LootTable table = banner.getLootTable();
+            if (table == null) {
+                itemId = doPull(banner, gachaInfo, pools);
+                itemData = GameData.getItemDataMap().get(itemId);
+            } else {
+                LootContext ctx = playerRecords.computeIfAbsent(player.getUid(), e -> new LootContext());
```
Comment (Birdulon) [inline src/main/java/emu/grasscutter/game/gacha/GachaManager.java]: This doesn't seem to be serialized, I haven't run this but I expect this to not save banner state between sessions. It also seems to be a single context (i.e. not per banner)?

File: src/main/java/emu/grasscutter/game/inventory/GameItem.java [hunk 2]
```diff
@@ -1,18 +1,6 @@
...
-
-import org.bson.types.ObjectId;
-
-import dev.morphia.annotations.Entity;
-import dev.morphia.annotations.Id;
-import dev.morphia.annotations.Indexed;
-import dev.morphia.annotations.PostLoad;
-import dev.morphia.annotations.Transient;
-
+import dev.morphia.annotations.*;
```
Comment (Birdulon) [inline src/main/java/emu/grasscutter/game/inventory/GameItem.java]: Is this purely a lint change on an unrelated file?

File: src/main/resources/defaults/data/Banners.json [hunk 3]
```diff
@@ -11,7 +11,8 @@
 		"endTime": 1924992000,
 		"sortId": 1000,
 		"fallbackItems4Pool1": [1006, 1014, 1015, 1020, 1021, 1023, 1024, 1025, 1027, 1031, 1032, 1034, 1036, 1039, 1043, 1044, 1045, 1048, 1053, 1055, 1056, 1064],
-		"weights5": [[1,75], [73,150], [90,10000]]
+		"weights5": [[1,75], [73,150], [90,10000]],
+        "lootTable": "gacha/StandardPool.json"
```
Comment (Birdulon) [inline src/main/resources/defaults/data/Banners.json]: This seems like a weird one if I understand it correctly, as the `lootTable` would override `weights5` and `fallbackItems4Pool1`.

**Review by Birdulon — COMMENTED** [review]
Leaving aside the current merge conflicts, this doesn't seem to be in a mergeable state.
As @Melledy mentioned earlier, the configuration is spread over too many `json` files, and while `MonsterDrop.json` being consolidated is a step in the right direction, it also shouldn't be one line as it makes diffs completely unusable.
Adding a bunch of files with the intention to delete them in a follow-up PR seems like a bad idea to me.

I'll have to look deeper into some other details on this PR later, and I do agree with the general ideas of it, but I'd prefer to sort things out a bit more before merging this.

File: src/main/java/emu/grasscutter/game/inventory/GameItem.java (hunk 2 above)
Comment (diauweb) [reply src/main/java/emu/grasscutter/game/inventory/GameItem.java]: IDEA did this without my agreement. I'll consider fix the history when everything else's set

File: src/main/java/emu/grasscutter/game/gacha/GachaManager.java (hunk 1 above)
Comment (diauweb) [reply src/main/java/emu/grasscutter/game/gacha/GachaManager.java]: Yes it is not serialized for now. I'll dive into it to find a way to serialize it properly. The general idea of this is instead of saving gacha pull counts in a separated table, we save it as a general data item. The states are meant to saved in a general data store and different banners just use different store keys to know their states.

File: src/main/java/emu/grasscutter/game/inventory/GameItem.java (hunk 2 above)
Comment (diauweb) [reply src/main/java/emu/grasscutter/game/inventory/GameItem.java]: > IDEA did this without my agreement. I'll consider fix the history when everything else's set

And all the 'bunch of files' mess. I'll force push to clean up to a mergable state after we reach an agreement.

File: src/main/resources/defaults/data/Banners.json (hunk 3 above)
Comment (diauweb) [reply src/main/resources/defaults/data/Banners.json]: Yes. if `lootTable` is set, all fields describing the output of the banner will be ignored since the loot table take the control of it

**Comment by diauweb** [comment]
> I'll have to look deeper into some other details on this PR later, and I do agree with the general ideas of it, but I'd prefer to sort things out a bit more before merging this.

It's a big, breaking change and do need more attention. It's important for me to receive reviews to make it good for everyone.

**Comment by eternalcomet** [comment]
object SceneObj_Area_Common_Property_Ani_Prop_BGM_01 should be add to file [ChestReward.json](https://github.com/Grasscutters/Grasscutter/pull/1379/files#diff-11401dbae7d39add289fd03859624299dc0b423777d1f36a1955faa5636f9324) ,because i see
`WARN:GadgetChest Could not found the handler of this type of Chests` when trying to pick up bgm.
表现为无法拾取旋曜玉帛

**Comment by diauweb** [comment]
> object SceneObj_Area_Common_Property_Ani_Prop_BGM_01 should be add to file [ChestReward.json](https://github.com/Grasscutters/Grasscutter/pull/1379/files#diff-11401dbae7d39add289fd03859624299dc0b423777d1f36a1955faa5636f9324) ,because i see `WARN:GadgetChest Could not found the handler of this type of Chests` when trying to pick up bgm. 表现为无法拾取旋曜玉帛

It's somehow unrelated to this PR. However you can solve this problem by adding your tables after this PR is merged

File: src/main/java/emu/grasscutter/game/gacha/GachaManager.java (hunk 1 above)
Comment (diauweb) [reply src/main/java/emu/grasscutter/game/gacha/GachaManager.java]: now the data is serialized

File: src/main/resources/defaults/data/Banners.json (hunk 3 above)
Comment (diauweb) [reply src/main/resources/defaults/data/Banners.json]: cleaned up unused fields

**Comment by eternalcomet** [comment]
why haven't it been merged yet?
does it work?

**Comment by Birdulon** [comment]
# Recent DMCA compliance push closed active PRs.

Apologies for taking so long to address loot tables, things keep popping up and the todo pile keeps growing :smiling_face_with_tear:
